PR: Add can-cast, broadcast-arrays and broadcast-to (#132)
data-apis/array-api: spec/API_specification/data_type_functions.md

A conforming implementation of the array API standard must provide and support the following data type functions.

<!-- NOTE: please keep the constants in alphabetical order -->

## Objects in API
(function-broadcast-arrays)=
### broadcast_arrays(*args, /)

Broadcast any number of arrays against each other.

#### Parameters

-   ***args**: _\[&lt;array&gt;\]_

    -   arrays to broadcast.

#### Returns

-   **out**: _\[&lt;array&gt;\]_

    -   a list of arrays where more than one element of a broadcasted array may refer to a single memory location.

(function-broadcast-to)=

> ### broadcast_to(array, /, shape)

Reviewer: ```suggestion
### broadcast_to(x, shape, /)
```

Was there a particular reason to make `shape` positional or keyword? We don't do this for the creation functions atm, and I am not convinced that the benefits of keyword support outweigh the need to support multiple ways of doing the same thing.

Reviewer: Also, `x` is used as the argument convention for a single array argument.

Author: > Was there a particular reason to make shape positional or keyword?

Nope, this was my mistake 😅

> Broadcast an array to a new shape.

Reviewer: ```suggestion
Broadcasts an array to a specified shape.
```

> -   A `ValueError` exception must be raised if the array is not compatible with the new shape according  to the semantics defined in {ref}`broadcasting`.

Reviewer: Moving this to below.

```suggestion
```

#### Parameters

> -   **array**: _&lt;array&gt;_

Reviewer: ```suggestion
-   **x**: _&lt;array&gt;_
```

    -   array to broadcast.

> -   **shape**: Tuple[int, …]

Reviewer: ```suggestion
-   **shape**: _Tuple\[int, ...]_
```

>     -   shape of the desired array.

Reviewer: ```suggestion
    -   array shape. Must be compatible with `x` (see {ref}`broadcasting`).
```

#### Returns

-   **out**: _&lt;array&gt;_

    -   an array with the given shape where more than one element may refer to a single memory location.

> 

Reviewer: ```suggestion

#### Raises

-   if the array is incompatible with the specified shape (see {ref}`broadcasting`).

```

(function-can-cast)=

> ### can_cast(from_, /, *, to)

Reviewer: ```suggestion
### can_cast(from, to, /)
```

`to` cannot be optional.

While `from` is a reserved keyword in Python, I don't believe we need to append a `_` suffix, as this argument is positional only.

> Determines if a type conversion is allowed under the rules described by {ref}`type-promotion`.

Reviewer: ```suggestion
Determines if one data type can be cast to another data type according {ref}`type-promotion` rules.
```

#### Parameters

-   **from_**: _Union\[ &lt;dtype&gt, &lt;array&gt;]_

>     -   data type, scalar or array to cast from.

Reviewer: ```suggestion
    -   input data type or array from which to cast.
```

> -   **to**: _&lt;dtype&gt;_

Reviewer: Not clear to me why NumPy, e.g., does not support both `dtype` and `array` here. This would allow

```python
result = can_cast(x1, x2)
```

where we can answer the question "Can `x1` be cast to the same dtype as `x2`?" without needing to explicitly type `x2.dtype`.

Author: I think we should add `array` in the spec, it makes sense to answer "Can `x1` be cast to the same dtype as `x2`?"

>     -   data type to cast to.

Reviewer: ```suggestion
    -   desired data type.
```

#### Returns

-   **out**: _bool_

>     -   True if the cast can occur according to the {ref}`type-promotion`.

Reviewer: ```suggestion
    -   `True` if the cast can occur according to {ref}`type-promotion` rules; otherwise, `False`.
```

> 

Reviewer: ```suggestion
```

<!-- NOTE: please keep the constants in alphabetical order -->

## Objects in API

> (broadcast-arrays)=

Reviewer 2: ```suggestion
(function-broadcast_arrays)=
```

### broadcast_arrays(args, /, *)

Broadcast any number of arrays against each other.

#### Parameters

> -   **args**: [_&lt;array&gt;_]

Reviewer 2: ```suggestion
-   **args**: _&lt;array&gt;_
```

> -   **args**: [_&lt;array&gt;_]

Reviewer 2: ```suggestion
-   **args**: _[&lt;array&gt;]_
```

    -   arrays to broadcast.

#### Returns

> -   **out**: [_&lt;array&gt;_]

Reviewer 2: ```suggestion
-   **out**: _[&lt;array&gt;]_
```

    -   readonly views on the original arrays. Furthermore, more than one element of a broadcasted array may refer to a single memory location.

> (broadcast-to)=

Reviewer 2: ```suggestion
(function-broadcast_to)=
```

> ### broadcast_to(array, /, shape, /)

Author: Ok! I'm still a little bit confused by the number of `/` in the signatures 😅

Broadcasts an array to a specified shape.

#### Parameters

-   **x**: _&lt;array&gt;_

    -   array to broadcast.

-   **shape**: _Tuple\[int, ...]_

    -   array shape. Must be compatible with `x` (see {ref}`broadcasting`).

#### Returns

-   **out**: _&lt;array&gt;_

    -   an array having a specified shape. Must have the same data type as `x`.

#### Raises

> 	-   if the array is incompatible with the specified shape (see {ref}`broadcasting`).

Reviewer: ```suggestion
-   if the array is incompatible with the specified shape (see {ref}`broadcasting`).
```

> (can-cast)=

Reviewer 2: ```suggestion
(function-can_cast)=
```

### can_cast(from_, /, to, /)

Determines if a type conversion is allowed under the rules described by {ref}`type-promotion`.

#### Parameters

> -   **from_**: _Union\[ &lt;dtype&gt ]_

Reviewer 3: There's either no need for the `Union` here, or it's missing `<array>, complex`.

Author: I completely missed them!

Reviewer 2: ```suggestion
-   **from_**: _Union\[ &lt;dtype&gt; ]_
```

>     -   data type, scalar or array to cast from.

Reviewer 3: This needs to be discussed - numpy allows Python scalars and arrays here. Other libraries may not. Would be nice to see the data on which libraries do what before making the decision.

Author: Given the docs numpy, jax and cupy seem to support scalars, arrays and dtypes. I'm going to add arrays for now and I'll check the usage of this function

Author: Edit: reviewing the python-record-api repo it seems that scalars and arrays are unused in the libraries we ran tests for. They are just using dtypes, should we just leave dtypes in the spec?

Reviewer 3: Thanks for digging up the data here. I agree, let's just leave `dtype`.

Reviewer: I think both dtypes and arrays should be supported here. This would be consistent with the `result_type` API which also supports both dtypes and arrays.

> -   **to**: _&lt;dtype&gt_

Reviewer 2: ```suggestion
-   **to**: _&lt;dtype&gt;_
```
